src/py-opentimelineio/opentimelineio/adapters/fcp_xml.py

```diff
@@ -619,12 +620,18 @@ def _derefed_element(self, element):
         Given an element, dereferences it by it's id attribute if needed. If
         the element has an id attribute and it's our first time encountering
         it, store the id.
+
+        Hiero and Nuke Studio generates xml files where a clip of a file into
+        two got the same file id for both files in the clips.
+        By adding a uuid4 to the elem_id name OpenTimelineIO parses all the
+        files correctly. This is only used during read and can not be seen
+        in the converted otio file.
         """
         if element is None:
             return element
 
         try:
-            elem_id = element.attrib["id"]
+            elem_id = "{}{}".format(element.attrib["id"], uuid.uuid4())
```

Comment:
This fix is problematic because it makes `elem_id` is unique every time the function is invoked - even if invoked twice with the exact same element. This renders `self._id_map` useless because when `setdefault` is called below, `elem_id` will always be unique and return the element passed in.

Unfortunately, this behavior breaks files using the `id` attribute as intended. This is also why the unittests are breaking - there are `MissingReferences` getting introduced in the place of `ExternalReferences` for empty `file` elements meant to be resolved to previous occurrences.

Here is the reference in the [FCP 7 XML specification](https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/Basics/Basics.html#//apple_ref/doc/uid/TP30001154-TPXREF126) providing some context around how the `id` attribute is meant to be used. 

In the sample data repo, there is an [example file](https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/tests/sample_data/premiere_example.xml) using the `id` behavior as designed. Here is the original declaration for `file-1`:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/tests/sample_data/premiere_example.xml#L67-L109

And here is a subsequent usage intended to resolve to the content of the first occurrence:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/tests/sample_data/premiere_example.xml#L436

If you search that file, you can see `file-1` referenced repeatedly throughout meant to re-use that original declaration.

Maybe the pragmatic solution would be to add some logic saying that if the element has children, update the cache for that `id` and return the element. If the element doesn't have children, use the current `setdefault` behavior. For the set of test EDLs I've seen, I think that should have the expected behavior.

Reply:
I'll try and dig into the code and see if I can find some other way to fix it.